### ISSUE TITLE
Bug (?): `PACKAGE_ROOT` not working one level up

### ISSUE THREAD

**Issue by jamesbraza** [astral-sh/uv#3271]
I have a `setuptools.build_meta` backend monorepo with a `pyproject.toml` package named `packagename` in a subfolder. I specify its dependency in a `requirements.txt` at the repo root: `package-name @ file://${PROJECT_ROOT}/packagename`

In `/Users/user/code/repo-root`, running `uv pip install --no-cache .` works fine.

Moving up one directory to `/Users/user/code` and running `uv pip install --no-cache ./repo` fails:

```none
error: Distribution not found at: file:///Users/user/code/packagename
```

Can we somehow know `PROJECT_ROOT` should be `./repo`, instead of `./` in this case?

---

I am using `uv 0.1.38 (0b23caa18 2024-04-24)`

**Comment by charliermarsh** [comment]
Hmm, right now `PROJECT_ROOT` is just the current working directory.

**Comment by charliermarsh** [comment]
Slightly tempted to set it to the directory of the containing file.

**Comment by jamesbraza** [comment]
Give in to your temptations 👿 🔥 🚀 🦾 😆

**Comment by zanieb** [comment]
Is that "breaking" ?

**Comment by charliermarsh** [comment]
I'd consider it breaking.

**Comment by zanieb** [comment]
Would we support `$PWD` too or whatever?

**Comment by charliermarsh** [comment]
I believe we already do.

**Comment by charliermarsh** [comment]
We support any env var, and then `PROJECT_ROOT` is special on top of those.

**Comment by charliermarsh** [comment]
Using the directory that contains the file definitely seems right for `pyproject.toml`. I could see it being less convenient for requirements.txt in some rare cases?

**Comment by zanieb** [comment]
Yeah the latter seems more problematic, e.g. `requirements/dev.txt` — but it seems quite correct for `pyproject.toml`.

I wonder if we should scan for `git` boundaries for `requirements.txt`...

**Comment by charliermarsh** [comment]
> Yeah the latter seems more problematic, e.g. requirements/dev.txt — but it seems quite correct for pyproject.toml.

Yeah. Is it even that bad though? It just means you need to make the path relative to the file rather than the root directory. But it would still work consistently regardless of where you invoked uv from, unlike the current implementation.

**Comment by zanieb** [comment]
I'm curious if anyone else will weigh in, but yeah having it work consistently seems like an improvement.

**Comment by jamesbraza** [comment]
Maybe `PACKAGE_ROOT` can have a fallback (or a configuration option `--package-root-is-not-cwd`) to maintain compatibility:

1. Looks in target directory (in this case, `./repo`)
2. Falls back on current working directory

That being said, imo the most straightforward/intuitive route is to support two environment variables:

```none
# Uses input uv command to resolve to ./repo/foo
file://${PROJECT_ROOT}/foo

# Always resolves to ./foo
file://${CURRENT_DIRECTORY}/foo
```

**Comment by charliermarsh** [comment]
For `${CURRENT_DIRECTORY}` you can likely already do `${PWD}`.

**Comment by charliermarsh** [comment]
We'll ship this in v0.3.0.

**Comment by ShapelessCat** [comment]
@jamesbraza In the title of this issue and in one of your comment, you used the `PACKAGE_ROOT`, not `PROJECT_ROOT`. Are those typos or because of some historical reason?

**Comment by jamesbraza** [comment]
I believe I used `PACKAGE_ROOT` and `PROJECT_ROOT` interchangeably in [this comment](https://github.com/astral-sh/uv/issues/3271#issuecomment-2079785370), both to refer to a path to the repo root containing the `pyproject.toml`.

I think `PROJECT_ROOT` makes more sense than `PACKAGE_ROOT`, given [workspaces](https://docs.astral.sh/uv/concepts/projects/workspaces/#workspace-sources) can have 2+ packages. Really `PROJECT_ROOT` is the path to the repo root

**Comment by m1cr0man** [comment]
Hi there. I just ran into this myself. I fully expected `PROJECT_ROOT` to refer to the directory containing the pyproject.toml. This is particularly problematic when using `uv run --script` in a shebang. If I call the script with the shebang from a working directory other than the project root (example `../myscript.py`), I get a `Distribution not found` error on the requirements utilising `PROJECT_ROOT` as the path is incorrect.

I understand there are concerns about this being a breaking change, but there is no workaround in this scenario. I cannot use `uv run --script` in shebangs and resolve the project root without introducing my own environment variable which I really do not want to do. With the behaviour corrected, any configurations relying on the old behaviour can simply switch to `${PWD}` as pointed out in previous comments.

**Comment by zanieb** [comment]
We'll need to drive https://github.com/astral-sh/uv/issues/13239 to consensus in order to reach a solution here, unfortunately. The pip compatibility issue is a major concern, on top of this being a breaking change for our existing users.

**Comment by m1cr0man** [comment]
I see. Are you looking for user feedback in that issue or is it more a matter for maintainers? I'm personally happy with what you have proposed.
